### PR TITLE
[MRG+1] Migrating selectors to use parsel

### PR DESCRIPTION
**Pull request by eliasdorneles**
Hey, folks!

So, here is my first stab at porting selectors to use [Parsel](https://github.com/scrapy/parsel).

~~I'm not very happy about the initialization of the `self._parser` attribute which is used to build the `LxmlDocument` instance for the response, that forced me to import `_ctgroup` from parsel.
I'd love to hear any suggestions about how to handle that.~~

What do you think?
Thank you!

### CODE REVIEW

File: scrapy/selector/unified.py [hunk 1]
```diff
@@ -47,111 +29,27 @@ def _response_from_text(text, st):
...
-        # set:difference
-        # set:has-same-node
-        # set:intersection
-        # set:leading
-        # set:trailing
-        "set": "http://exslt.org/sets"
-    }
-    _lxml_smart_strings = False
+        # supporting legacy _root argument
+        root = kwargs.get('_root', root)
```
Comment (eliasdorneles) [inline scrapy/selector/unified.py]: I'm not sure if this is really needed, since the root argument seems to be only used here: https://github.com/scrapy/parsel/blob/master/parsel/unified.py#L96

Do you think we could get rid of this?

File: scrapy/selector/unified.py [hunk 2]
```diff
@@ -47,111 +29,27 @@ def _response_from_text(text, st):
...
         self._parser = _ctgroup[st]['_parser']
-        self._csstranslator = _ctgroup[st]['_csstranslator']
-        self._tostring_method = _ctgroup[st]['_tostring_method']
 
         if text is not None:
             response = _response_from_text(text, st)
 
         if response is not None:
-            _root = LxmlDocument(response, self._parser)
+            root = LxmlDocument(response, self._parser)
```
Comment (dangra) [inline scrapy/selector/unified.py]: I think we can drop LxmlDocument, it is a cache of parsed dom keyedb by Response instance, it is rarely needed now that we have a shortcut to selector from response. The old way of instanciating selectors manually is not encouraged since a few versions.

File: scrapy/selector/unified.py [hunk 3]
```diff
@@ -47,111 +29,27 @@ def _response_from_text(text, st):
...
-                    el.attrib[an.split('}', 1)[1]] = el.attrib.pop(an)
-
-    def __nonzero__(self):
-        return bool(self.extract())
-
-    def __str__(self):
-        data = repr(self.extract()[:40])
-        return "<%s xpath=%r data=%s>" % (type(self).__name__, self._expr, data)
-    __repr__ = __str__
+        text = response.body_as_unicode() if response else None
```
Comment (dangra) [inline scrapy/selector/unified.py]: I would check for `if response is None` just in case.

File: scrapy/selector/unified.py (hunk 1 above)
Comment (dangra) [inline scrapy/selector/unified.py]: I'd leave it and log a warning about change of parameter name. Same goes if we decide to promote self._root to self.root.

File: scrapy/selector/unified.py [hunk 4]
```diff
@@ -47,111 +29,27 @@ def _response_from_text(text, st):
...
-        "set": "http://exslt.org/sets"
-    }
-    _lxml_smart_strings = False
+        # supporting legacy _root argument
+        root = kwargs.get('_root', root)
 
-    def __init__(self, response=None, text=None, type=None, namespaces=None,
-                 _root=None, _expr=None):
-        self.type = st = _st(response, type or self._default_type)
         self._parser = _ctgroup[st]['_parser']
```
Comment (dangra) [inline scrapy/selector/unified.py]: _ctgroup deserves a better (public) name

File: scrapy/selector/csstranslator.py [hunk 5]
```diff
@@ -1,88 +1,6 @@
...
-
-class ScrapyHTMLTranslator(TranslatorMixin, HTMLTranslator):
-    pass
-
+from parsel.csstranslator import (
+    ScrapyXPathExpr,
+    TranslatorMixin,
+    ScrapyGenericTranslator,
+    ScrapyHTMLTranslator
+)
```
Comment (dangra) [inline scrapy/selector/csstranslator.py]: I would go for deprecation on all this classes, instruct users to import from parsel directly.

File: scrapy/selector/unified.py (hunk 4 above)
Comment (eliasdorneles) [inline scrapy/selector/unified.py]: what about adding a property to parsel's Selector that returns the proper object according to self.type?

```
@property
def _parser(self):
    return _ctgroup[self.type]['_parser']
```

File: scrapy/selector/unified.py (hunk 3 above)
Comment (eliasdorneles) [inline scrapy/selector/unified.py]: hm, do response objects can have non-truthy values?

Comment (dangra) [inline scrapy/selector/unified.py]: No, but truth of some objects can become expensive if someone decides to pass a cleverish response. It happened before with selectors checks calling overrided `__nonzero__`

File: scrapy/selector/unified.py (hunk 2 above)
Comment (eliasdorneles) [inline scrapy/selector/unified.py]: I like this -- I'll be able to get rid of the _ctgroup import and code will be cleaner. :D

File: scrapy/selector/unified.py (hunk 4 above)
Comment (dangra) [inline scrapy/selector/unified.py]: actually, I think we shouldn't be setting _parser in this constructor. Instead call `super().__init__(text=response.body_as_unicode(), type=st)` and let Parsel selector do the parsing based on type.

File: scrapy/selector/unified.py (hunk 2 above)
Comment (dangra) [inline scrapy/selector/unified.py]: :+1:

File: scrapy/selector/unified.py (hunk 4 above)
Comment (eliasdorneles) [inline scrapy/selector/unified.py]: yeah, I was only setting it here because I needed it for `LxmlDocument`, thought we were keeping it.
I'll update the PR tomorrow.

Comment (dangra) [inline scrapy/selector/unified.py]: :ok:

File: scrapy/selector/csstranslator.py (hunk 5 above)
Comment (eliasdorneles) [inline scrapy/selector/csstranslator.py]: hm, I suppose I should rename those in Parsel too. `ParselGenericTranslator`, etc.
this will need a new release of parsel, but I think it's worthy, right?

Comment (dangra) [inline scrapy/selector/csstranslator.py]: no need to rename in parsel, you can do like https://github.com/scrapy/scrapy/blob/311293ffdc63892bd5ab8494310529a6da0f5b62/scrapy/settings/__init__.py#L177-L197

Comment (dangra) [inline scrapy/selector/csstranslator.py]:
```
from parsel.csstranslator import ScrapyXPathExpr

ScrapyXPathExpr = create_deprecated_class("ScrapyXPathExpr", ScrapyXPathExpr, new_class_path='parsel.csstranslator.ScrapyXPathExpr')
```

Comment (eliasdorneles) [inline scrapy/selector/csstranslator.py]: I understand, but my question is more like: "should Parsel have classes with Scrapy on their names?"

Comment (dangra) [inline scrapy/selector/csstranslator.py]: oh you're right! no for sure but having Parsel prefix does't make sense neither.

Comment (dangra) [inline scrapy/selector/csstranslator.py]: to your original question, let's remove prefixes from parsel and do a new release

Comment (eliasdorneles) [inline scrapy/selector/csstranslator.py]: :+1:

File: tests/test_selector.py [hunk 6]
```diff
@@ -37,6 +39,15 @@ def test_simple_selection(self):
         self.assertEqual([x.extract() for x in sel.xpath("concat(//input[@name='a']/@value, //input[@name='b']/@value)")],
                          [u'12'])
 
+    @mock.patch('scrapy.selector.unified.warnings')
+    def test_deprecated_root_argument(self, warnings):
+        root = etree.fromstring(u'<html/>')
+        sel = self.sscls(_root=root)
+        self.assertEqual(root, sel._root)
```
Comment (dangra) [inline tests/test_selector.py]: I think it should be `self.assertIs()`

**Comment by kmike** [comment]
Without LxmlDocument cache some parts of Scrapy can become less efficient. For example, LxmlLinkExtractor instantiates Selector(response), so for users which use both link extractors and selectors response will be parsed twice. Maybe there are other cases like this (maybe in user's components), I'm not sure. If we proceed with cache removal I think we should at least fix Scrapy built-in components.

**Comment by dangra** [comment]
> If we proceed with cache removal I think we should at least fix Scrapy built-in components.

do you mean replacing calls like `html = Selector(response)` by `html = response.selector`? seems quite easy.

**Comment by kmike** [comment]
> do you mean replacing calls like html = Selector(response) by html = response.selector? seems quite easy.

yep, that's it

**Comment by eliasdorneles** [comment]
should I make that part of this PR?

**Comment by dangra** [comment]
for the code that force the selector `type` que just let it be.

```
~/src/scrapy$ ack 'Selector\(' scrapy/
scrapy/http/response/text.py:106:            self._cached_selector = Selector(self)
scrapy/selector/unified.py:50:class Selector(object_ref):
scrapy/spiders/feed.py:72:            selector = Selector(response, type='xml')
scrapy/spiders/feed.py:76:            selector = Selector(response, type='html')
scrapy/linkextractors/sgml.py:130:            sel = Selector(response)
scrapy/linkextractors/lxmlhtml.py:68:        html = Selector(response)
scrapy/linkextractors/lxmlhtml.py:98:        html = Selector(response)
scrapy/utils/iterators.py:40:        yield Selector(text=nodetext, type='xml').xpath('//' + nodename)[0]
scrapy/utils/iterators.py:52:        xs = Selector(text=nodetext, type='xml')
```

**Comment by dangra** [comment]
@eliasdorneles yes, I think so if it pass tests with that simple change.

**Comment by eliasdorneles** [comment]
@dangra talking about tests, any idea why I'm getting failures on `test_squeues.py` there? :point_up:

**Comment by dangra** [comment]
> talking about tests, any idea why I'm getting failures on test_squeues.py there? 
>  Hide all checks

no, it started to happen on a previous but very unlikely commit build https://travis-ci.org/scrapy/scrapy/jobs/74296461

**Comment by dangra** [comment]
if it is passing for "precise" then it is probably a bug introduced by some dependency upgrade 
image

Twisted 15.3.0 was released yesterday for example

**Comment by dangra** [comment]
It doesn't make sense to me

```
>>> import sys, pickle, cPickle
>>> sys.version
'2.7.10 (default, May 26 2015, 04:16:29) \n[GCC 5.1.0]'
>>> cPickle.dumps((lambda x: x), protocol=2)
---------------------------------------------------------------------------
PicklingError                             Traceback (most recent call last)
<ipython-input-10-e076fdfef0cc> in <module>()
----> 1 cPickle.dumps((lambda x: x), protocol=2)

PicklingError: Can't pickle <type 'function'>: attribute lookup __builtin__.function failed
>>> pickle.dumps((lambda x: x), protocol=2)
---------------------------------------------------------------------------
PicklingError                             Traceback (most recent call last)
<ipython-input-11-fb0cc15a50f8> in <module>()
----> 1 pickle.dumps((lambda x: x), protocol=2)

/usr/lib64/python2.7/pickle.pyc in dumps(obj, protocol)
   1372 def dumps(obj, protocol=None):
   1373     file = StringIO()
-> 1374     Pickler(file, protocol).dump(obj)
   1375     return file.getvalue()
   1376 

/usr/lib64/python2.7/pickle.pyc in dump(self, obj)
    222         if self.proto >= 2:
    223             self.write(PROTO + chr(self.proto))
--> 224         self.save(obj)
    225         self.write(STOP)
    226 

/usr/lib64/python2.7/pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288 

/usr/lib64/python2.7/pickle.pyc in save_global(self, obj, name, pack)
    746             raise PicklingError(
    747                 "Can't pickle %r: it's not found as %s.%s" %
--> 748                 (obj, module, name))
    749         else:
    750             if klass is not obj:

PicklingError: Can't pickle <function <lambda> at 0x7f48660f6848>: it's not found as __main__.<lambda>
>>> 
```

**Comment by eliasdorneles** [comment]
@dangra it's definitely weird, but I tested locally fixing Twisted version to 15.2.1 and it passed.
no idea why, though.

**Comment by kmike** [comment]
On Travis it is failing with AssertionError, but we're expecting ValueError in test. I have no idea how can it depend on installed requirements. For me this test passes locally with twisted 15.3.0.

**Comment by dangra** [comment]
@kmike it fails if you run the whole testsuite but pass if you run just that test file

**Comment by eliasdorneles** [comment]
okay, this is weird.

if I run (with last Twisted) just that test with: `tox -r -e py27 -- tests/test_squeues.py`, it passes.

but if I run the whole suite with `tox -r -e py27`, it fails.

ideas?

**Comment by dangra** [comment]
running all tests with twisted 15.2.0 passes for me

**Comment by eliasdorneles** [comment]
yeah, the same with 15.2.1 for me.

**Comment by dangra** [comment]
It fails for 15.3.0 using `$ tox -e py27 tests/test_crawl.py tests/test_squeues.py`

**Comment by kmike** [comment]
I think it can be related to https://github.com/twisted/twisted/commit/a8d8a0c89dc530554c768739e83d7a852cb17f22 - see [this](https://github.com/twisted/twisted/blob/a8d8a0c89dc530554c768739e83d7a852cb17f22/twisted/persisted/styles.py#L140) line.

**Comment by dangra** [comment]
@kmike that's for sure.

**Comment by eliasdorneles** [comment]
@dangra @kmike so, I've made the changes to use `response.selector` in link extractors.

The only place still using LxmlDocument is here: https://github.com/scrapy/scrapy/blob/master/scrapy/http/request/form.py#L60

I get a bunch of test failures if I try to replace that for `response.selector`, so I suppose it's not safe to make the assumption all responses there will have a selector for it -- should we keep it there?

**Comment by dangra** [comment]
> The only place still using LxmlDocument is here: https://github.com/scrapy/scrapy/blob/master/scrapy/http/request/form.py#L60

@elias replace it by the corr espondent direct call to lxml parser following re-encoding best practices from parsel (i.e.: body_as_unicode().encode('utf8'))

**Comment by eliasdorneles** [comment]
@dangra okay, finally got rid of it. :) :tada: 
should I rebase and/or squash the commits?

File: scrapy/http/request/form.py [hunk 7]
```diff
@@ -54,10 +55,15 @@ def _urlencode(seq, enc):
     return urlencode(values, doseq=1)
 
 
+def _create_parser_from_response(response, parser_cls):
```
Comment (kmike) [inline scrapy/http/request/form.py]: +1 to expose this function in parsel instead of duplicating it

Comment (dangra) [inline scrapy/http/request/form.py]: @kmike what do you think about promoting `Selector._root` to `Selector.root` in parsel library and use an `extended-html` parser for `lxml.html.HTMLParser` ?

Comment (kmike) [inline scrapy/http/request/form.py]: Sorry, I feel dumb - could you please explain what is `extended-html` and why is it different from `html`?

Promoting `_root` to `root` sounds good, I had to use `_root` more than once.

Comment (eliasdorneles) [inline scrapy/http/request/form.py]: @kmike in practice, it would mean using [`lxml.html.HTMLParser`](http://lxml.de/api/lxml.html.HTMLParser-class.html) instead of [lxml.etree.HTMLParser](http://lxml.de/api/lxml.etree.HTMLParser-class.html) (which is extended by the former).

I'll promote `_root` in parsel, and add a deprecated attribute for `_root` in Scrapy's subclass.

Comment (dangra) [inline scrapy/http/request/form.py]: > Sorry, I feel dumb - could you please explain what is extended-html and why is it different from html?

FormRequest uses `lxml.html.HTMLParser` which is different from `type=html` in `Selector` class which uses `lxml.etree.HTMLParser`. 

The former provides extra methods to query html in a python way but it is slower than the later.

Comment (kmike) [inline scrapy/http/request/form.py]: why can't we use lxml.html.HTMLParser by default?

Comment (dangra) [inline scrapy/http/request/form.py]: Back in the future https://github.com/scrapy/scrapy/pull/471#issuecomment-34008359

File: scrapy/selector/unified.py [hunk 8]
```diff
@@ -47,113 +26,34 @@ def _response_from_text(text, st):
...
+    def __init__(self, response=None, text=None, type=None, root=None, _root=None, **kwargs):
+        st = _st(response, type or self._default_type)
 
-    def __init__(self, response=None, text=None, type=None, namespaces=None,
-                 _root=None, _expr=None):
-        self.type = st = _st(response, type or self._default_type)
-        self._parser = _ctgroup[st]['_parser']
-        self._csstranslator = _ctgroup[st]['_csstranslator']
-        self._tostring_method = _ctgroup[st]['_tostring_method']
+        if root is None and _root is not None:
```
Comment (eliasdorneles) [inline scrapy/selector/unified.py]: Okay, so this is the only bit that have me a tiny bit anxious.

Is it okay to ignore `_root` when we get a `root` argument (assuming `root` precedence will speak for itself), or should we always warn whenever there is a value for `_root`?

**Comment by eliasdorneles** [comment]
Alright, updated parsel and the PR, I think I'm almost done now.
What do you think?

File: scrapy/selector/unified.py (hunk 8 above)
Comment (kmike) [inline scrapy/selector/unified.py]: for me both options look fine.

File: scrapy/linkextractors/sgml.py [hunk 9]
```diff
@@ -127,7 +127,7 @@ def __init__(self, allow=(), deny=(), allow_domains=(), deny_domains=(), restric
     def extract_links(self, response):
         base_url = None
         if self.restrict_xpaths:
-            sel = Selector(response)
+            sel = response.selector
```
Comment (kmike) [inline scrapy/linkextractors/sgml.py]: maybe we can drop sel variable and just use response.xpath?

File: scrapy/selector/unified.py (hunk 8 above)
Comment (kmike) [inline scrapy/selector/unified.py]: after thinking a bit more about it I think we should always issue a warning :)

File: tests/test_selector.py [hunk 10]
```diff
@@ -37,6 +39,15 @@ def test_simple_selection(self):
...
                          [u'12'])
 
+    @mock.patch('scrapy.selector.unified.warnings')
+    def test_deprecated_root_argument(self, warnings):
+        root = etree.fromstring(u'<html/>')
+        sel = self.sscls(_root=root)
+        self.assertIs(root, sel.root)
+        warnings.warn.assert_called_once_with(
+            'Argument `_root` is deprecated, use `root` instead',
+            mock.ANY, stacklevel=2)
```
Comment (kmike) [inline tests/test_selector.py]: Is it possible to test it without mocks, like described in [docs](https://docs.python.org/2/library/warnings.html#testing-warnings)?

File: scrapy/selector/unified.py (hunk 8 above)
Comment (eliasdorneles) [inline scrapy/selector/unified.py]: :+1: , I'll update the PR addressing this and other comments too.

**Comment by eliasdorneles** [comment]
Updated addressing @kmike's comments and rebased on top of current master.

**Comment by dangra** [comment]
tests/test_selector_csstranslator.py can be removed or replaced by tests to assert it warns.

test/selector.py can be simplified due to most tests moved to parsel, we can left flavor detection and any other response based specific test case.

after that I think it will just work on python 3 too and tests/py3-ignores.txt updated.

**Comment by eliasdorneles** [comment]
@dangra done!
I also left the first simple test for selector to serve as base case.

File: tests/test_selector.py [hunk 11]
```diff
@@ -7,22 +7,21 @@
...
 class SelectorTestCase(unittest.TestCase):
 
-    sscls = Selector
-
     def test_simple_selection(self):
         """Simple selector tests"""
-        body = "<p><input name='a'value='1'/><input name='b'value='2'/></p>"
-        response = TextResponse(url="http://example.com", body=body)
-        sel = self.sscls(response)
+        body = u"<p><input name='a'value='1'/><input name='b'value='2'/></p>"
```
Comment (dangra) [inline tests/test_selector.py]: it is more direct to use `b` prefix and avoid encoding.

File: tests/test_selector.py [hunk 12]
```diff
@@ -37,234 +36,34 @@ def test_simple_selection(self):
...
+            root = etree.fromstring(u'<html/>')
+            sel = Selector(_root=_root, root=root)
+            self.assertIs(root, sel.root)
+            self.assertIn('Ignoring deprecated `_root` argument', str(w[-1].message))
 
     def test_flavor_detection(self):
-        text = '<div><img src="a.jpg"><p>Hello</div>'
-        sel = self.sscls(XmlResponse('http://example.com', body=text))
+        text = u'<div><img src="a.jpg"><p>Hello</div>'
+        sel = Selector(XmlResponse('http://example.com', body=text, encoding='utf-8'))
```
Comment (dangra) [inline tests/test_selector.py]: same thing about passing bytes as body instead of uncode.

File: tests/test_selector.py [hunk 13]
```diff
@@ -280,136 +79,28 @@ def test_http_header_encoding_precedence(self):
...
-        r1 = TextResponse('http://www.example.com', \
-                          body='<root>pre\x00post</root>', \
-                          encoding='utf-8')
-        self.sscls(r1).xpath('//text()').extract()
-
     def test_badly_encoded_body(self):
         # \xe9 alone isn't valid utf8 sequence
         r1 = TextResponse('http://www.example.com', \
-                          body='<html><p>an Jos\xe9 de</p><html>', \
+                          body=u'<html><p>an Jos\xe9 de</p><html>', \
```
Comment (dangra) [inline tests/test_selector.py]: the body must be bytes for this test case, it is a badly encoded utf8 sequence.

File: scrapy/selector/unified.py [hunk 14]
```diff
@@ -47,113 +26,37 @@ def _response_from_text(text, st):
...
+                root = _root
+            else:
+                warnings.warn("Ignoring deprecated `_root` argument, using provided `root`")
 
         if text is not None:
             response = _response_from_text(text, st)
 
         if response is not None:
-            _root = LxmlDocument(response, self._parser)
+            text = response.body_as_unicode()
```
Comment (dangra) [inline scrapy/selector/unified.py]: now that parsel can handle base_url, let's be full backward compatible: `kwargs.setdefault('base_url', response.url)`

Comment (eliasdorneles) [inline scrapy/selector/unified.py]: nice! :+1:

**Comment by eliasdorneles** [comment]
@dangra updated and rebased on top of current master :)

**Comment by dangra** [comment]
great work @elias!

**Comment by dangra** [comment]
@kmike feel free to merge when you are happy

File: scrapy/selector/unified.py [hunk 15]
```diff
@@ -161,43 +65,3 @@ def select(self, xpath):
...
-    def extract_unquoted(self):
-        return [x.extract_unquoted() for x in self]
-
-    @deprecated(use_instead='.xpath()')
-    def x(self, xpath):
-        return self.select(xpath)
-
-    @deprecated(use_instead='.xpath()')
-    def select(self, xpath):
-        return self.xpath(xpath)
```
Comment (kmike) [inline scrapy/selector/unified.py]: I think it is fine to remove these deprecated methods, but we should not forget to mention it in the release notes.

Comment (dangra) [inline scrapy/selector/unified.py]: I prefer to do that in another PR, this one is about migrating to use parsel lib while retaining current functionality

Comment (kmike) [inline scrapy/selector/unified.py]: Do you mean we should add backwards compatibility shims for these methods in this PR?

Comment (eliasdorneles) [inline scrapy/selector/unified.py]: woo, yeah, we're missing the shims for these methods! :O

Comment (dangra) [inline scrapy/selector/unified.py]: Yes, I think so. We are keeping them for `Selector` class

 It actually means that `SelectorList` must be a class attribute of `Selector` class

Comment (eliasdorneles) [inline scrapy/selector/unified.py]: right, so it can be defined by subclasses.
I'll change this in parsel, do a release and update this PR.

Comment (dangra) [inline scrapy/selector/unified.py]: Related https://github.com/scrapy/parsel/pull/11

Comment (eliasdorneles) [inline scrapy/selector/unified.py]: alright, shim added!
we're getting closer, guys! :D

**Comment by kmike** [comment]
Great job @eliasdorneles! Also, thanks @umrashrf for getting the ball rolling.
